Document sw lifetime workaround (#7792)
GoogleChrome/developer.chrome.com: site/en/docs/extensions/migrating/to-service-workers/index.md

});
```
{% endCompare %}

## Keeping a service worker alive

Service workers are by definition event-driven and will terminate on inactivity. This way Chrome can optimize performance and memory consumption of your extension. Learn more in our [service worker lifecycle documentation](/docs/extensions/mv3/service_workers/service-worker-lifecycle/#idle-shutdown). Exceptional cases might require additional measures to ensure that a service worker stays alive for a longer time.

### Keeping a service worker alive until a long-running operation is finished

When executing a long running operation that does not call any extension APIs in a service worker, the service worker might get shut down while the operation is running. Examples for when this might be needed are:

> * A [`fetch` request](https://developer.mozilla.org/docs/Web/API/Fetch_API) potentially taking longer than longer than 5 minutes (e.g. a large download on a potentially poor connection). 

Reviewer: I know the next sentence leaves the digits intact. The standard rules around English numbers are just weird that way.

* A complex asynchronous calculation taking more than 30 seconds.

To extend the service worker lifetime in these cases, you can periodically call a trivial extension API to reset the timeout counter. 
Please note, that this is only reserved for exceptional cases and in most situations there is usually a better, platform idiomatic, way to achieve the same result.

> The following code snippet shows a `waitUntil` helper function that will keep your service worker alive until a given promise resolves:

Reviewer: ```suggestion
The following example shows a `waitUntil()` helper function that keeps your service worker alive until a given promise resolves:
```

```js
async function waitUntil(promise) = {
  const keepAlive = setInterval(chrome.runtime.getPlatformInfo, 25 * 1000);
  try {
    await promise;
  } finally {
    clearInterval(keepAlive);
  }
}

waitUntil(someExpensiveCalculation());
```

{% Aside 'important' %}

> An official API similar to `waitUntil` is currently being discussed in the WECG. For more detail, see the [discussion on GitHub](https://github.com/w3c/webextensions/issues/416).

Reviewer: ```suggestion
An official API similar to `waitUntil()` is currently being discussed in the WECG. For more detail, see the [discussion on GitHub](https://github.com/w3c/webextensions/issues/416).
```

> An official API simmilar to `waitUntil` is currently being discussed in the WECG. For more detail, see the [discussion on Github](https://github.com/w3c/webextensions/issues/416).

Reviewer 2: ```suggestion
An official API similar to `waitUntil` is currently being discussed in the WECG. For more detail, see the [discussion on GitHub](https://github.com/w3c/webextensions/issues/416).
```

{% endAside %}

> ### Keeping a service worker alive continously

Reviewer 2: ```suggestion
### Keeping a service worker alive continuously
```

> ### Keeping a service worker alive continuously

Reviewer: ```suggestion
### Keep a service worker alive continuously
```

> In rare cases, it is necessary to extend the lifetime indefinitely. Enterprise and EDU are the biggest use cases, and we specifically allow this there, but we do not support this in general. In these exceptional circumstances, keeping a service worker alive can be achieved by periodically calling a trivial extension API. It is important to note that this recommendation only applies to extensions running on managed devices for enterprise or EDU use cases. It is not allowed in other cases and the Chrome extension team reserves the right to take action against those extensions in the future. 

Reviewer 2: ```suggestion
In rare cases, it is necessary to extend the lifetime indefinitely. We have identified enterprise and EDU as the biggest use cases, and we specifically allow this there, but we do not support this in general. In these exceptional circumstances, keeping a service worker alive can be achieved by periodically calling a trivial extension API. It is important to note that this recommendation only applies to extensions running on managed devices for enterprise or EDU use cases. It is not allowed in other cases and the Chrome extension team reserves the right to take action against those extensions in the future. 
```

Reviewer 2: Alternative suggestion:

```suggestion
It is sometimes desirable to extend the lifetime indefinitely. For extensions intended to be used
for enterprise or EDU use cases on managed devices, keeping a service worker alive can be achieved
by periodically calling a trivial extension API. Beyond this, we recommend the previous approach and
may take action in the future on extensions keeping their service worker alive after any work has
finished.
```

> In rare cases, it is necessary to extend the lifetime indefinitely. We have identified enterprise and EDU as the biggest use cases, and we specifically allow this there, but we do not support this in general. In these exceptional circumstances, keeping a service worker alive can be achieved by periodically calling a trivial extension API. It is important to note that this recommendation only applies to extensions running on managed devices for enterprise or EDU use cases. It is not allowed in other cases and the Chrome extension team reserves the right to take action against those extensions in the future. 

Reviewer: ```suggestion
In rare cases, it is necessary to extend the lifetime indefinitely. We have identified enterprise and education as the biggest use cases, and we specifically allow this there, but we do not support this in general. In these exceptional circumstances, keeping a service worker alive can be achieved by periodically calling a trivial extension API. It is important to note that this recommendation only applies to extensions running on managed devices for enterprise or education use cases. It is not allowed in other cases and the Chrome extension team reserves the right to take action against those extensions in the future. 
```

> Use the following code snippet to keep your service worker alive as long as your :

Reviewer 2: ```suggestion
Use the following code snippet to keep your service worker alive:
```

```js
/**
 * Tracks when a service worker was last alive and extends the service worker
 * lifetime by writing the current time to extension storage every 20 seconds.

>  * You should still prepare for unexpected termination - for example, if the

Reviewer: This seems like something you would cover in an intro paragraph.

Author: Yeah, but as it's intended to be copy pasted I thought it's good to leave this in the code.

Reviewer 2: To add to this, we expect people to copy paste this without looking at it so adding some detailed comments is a way to hopefully make them read it before they commit it.

Reviewer: If developers aren't going to read it, I don't think where you put it makes much of a differences. Frankly, the comments as they are now, don't give me much of a sense of what this is doing. Writing needs to tell me what's going on. A proper overview, which would contain much of the comment material would do that. 

In other words, the explanations are piecemeal and don't explain the big picture. That big picture would have to include enough of the piecemeal information to make the many of the code comments redundant.

Author: This looks fine as code comments. These comments are not for developers to be read, but to be copy pasted into the code base.
